concrete-core/src/specification/engines/glwe_ciphertext_discarding_relinearisation_computation.rs

```diff
@@ -0,0 +1,65 @@
+use super::engine_error;
+use crate::prelude::GlweCiphertext32;
+use crate::specification::engines::AbstractEngine;
+use crate::specification::entities::markers::GlweRelinearisationKeyKind;
+use crate::specification::entities::{GlweCiphertextEntity, GlweRelinearisationKeyEntity};
+
+engine_error! {
+    GlweCiphertextDiscardingRelinearisationError for GlweCiphertextDiscardingRelinearisationEngine @
+    PolynomialSizeMismatch => "The polynomial size of the input and output GLWE ciphertexts must be\
+     the same.",
+    InputGlweDimensionMismatch => "The GLWE dimension of the input ciphertext and key must be the\
+     same.",
+    OutputGlweDimensionMismatch => "The GLWE dimension of the output ciphertext is incorrect"
+}
+
+impl<EngineError: std::error::Error> GlweCiphertextDiscardingRelinearisationError<EngineError> {
+    pub fn perform_generic_checks<InputKey, InputCiphertext, OutputCiphertext>(
+        input_key: &InputKey,
+        input_ct: &InputCiphertext,
+        output: &OutputCiphertext,
+    ) -> Result<(), Self>
+    where
+        // TODO: do we need the key as input here?
+        InputKey: GlweRelinearisationKeyEntity,
+        InputCiphertext: GlweCiphertextEntity,
+        OutputCiphertext: GlweCiphertextEntity,
+    {
+        if input_ct.polynomial_size().0 != input_key.polynomial_size().0
+            || input_ct.polynomial_size().0 != output.polynomial_size().0
+        {
+            return Err(Self::PolynomialSizeMismatch);
+        }
+        if input_ct.glwe_dimension().0 != input_key.glwe_dimension().0 {
+            return Err(Self::InputGlweDimensionMismatch);
+        }
+        if input_ct.glwe_dimension().0
+            != 1 / 2 * output.glwe_dimension().0 * (3 + output.glwe_dimension().0)
+        {
+            return Err(Self::OutputGlweDimensionMismatch);
+        }
+        Ok(())
+    }
+}
+
+pub trait GlweCiphertextDiscardingTensorProductEngine<Inputkey, InputCiphertext, OutputCiphertext>:
+    AbstractEngine
+{
+    fn discard_relinearise_glwe_ciphertext(
+        &mut self,
+        input_key: GlweRelinearisationKeyKind,
```

Comment:
should be `InputKey`
shouldn't have any `GlweCiphertext32` in spec files, only generic type parameters, e.g. here we should use `InputCiphertext`